internal/component/discovery/target.go

```diff
@@ -0,0 +1,370 @@
+	})
+}
+
+func (t Target) HashLabelsWithPredicate(pred func(key string) bool) uint64 {
+	// For hash to be deterministic, we need labels order to be deterministic too. Figure this out first.
+	labelsInOrder := stringSlicesPool.Get().([]string)
+	defer stringSlicesPool.Put(labelsInOrder[:])
+	t.ForEachLabel(func(key string, value string) bool {
+		if pred(key) {
+			labelsInOrder = append(labelsInOrder, key)
+		}
+		return true
+	})
+	slices.Sort(labelsInOrder)
+	return t.hashLabelsInOrder(labelsInOrder)
+}
+
+func (t Target) groupLabelsHash() uint64 {
+	// For hash to be deterministic, we need labels order to be deterministic too. Figure this out first.
+	labelsInOrder := stringSlicesPool.Get().([]string)
+	defer stringSlicesPool.Put(labelsInOrder[:])
+
+	for name := range t.group {
+		labelsInOrder = append(labelsInOrder, string(name))
+	}
+	slices.Sort(labelsInOrder)
+	return t.hashLabelsInOrder(labelsInOrder)
+}
+
+func (t Target) hashLabelsInOrder(order []string) uint64 {
+	// This optimisation is adapted from prometheus/model/labels.
+	// Use xxhash.Sum64(b) for fast path as it's faster.
+	b := make([]byte, 0, 1024)
+	mustGet := func(label string) string {
+		val, ok := t.Get(label)
+		if !ok { // should never happen as Target is immutable
+			panic("label concurrently modified - this is a bug - please report an issue")
+		}
+		return val
+	}
+
+	for i, key := range order {
+		value := mustGet(key)
+		if len(b)+len(key)+len(value)+2 >= cap(b) {
+			// If labels entry is 1KB+ do not allocate whole entry.
+			h := xxhash.New()
+			_, _ = h.Write(b)
+			for _, key := range order[i:] {
+				_, _ = h.WriteString(key)
+				_, _ = h.Write(seps)
+				_, _ = h.WriteString(mustGet(key))
+				_, _ = h.Write(seps)
+			}
+			return h.Sum64()
+		}
+
+		b = append(b, key...)
+		b = append(b, seps[0])
+		b = append(b, value...)
+		b = append(b, seps[0])
+	}
+	return xxhash.Sum64(b)
+}
+
+var (
+	// used in tests to simulate hash conflicts
+	labelSetEqualsFn = func(l1, l2 commonlabels.LabelSet) bool { return &l1 == &l2 || l1.Equal(l2) }
+	stringSlicesPool = sync.Pool{New: func() interface{} { return make([]string, 0, 20) }}
+)
+
+func ComponentTargetsToPromTargetGroups(jobName string, tgs []Target) map[string][]*targetgroup.Group {
+	targetIndWithCommonGroupLabels := map[uint64][]int{} // target group hash --> index of target in tgs array
+	for ind, t := range tgs {
+		fp := t.groupLabelsHash()
+		targetIndWithCommonGroupLabels[fp] = append(targetIndWithCommonGroupLabels[fp], ind)
+	}
+
+	// Sort by hash to get deterministic order
+	sortedKeys := maps.Keys(targetIndWithCommonGroupLabels)
+	slices.Sort(sortedKeys)
+
+	allGroups := make([]*targetgroup.Group, 0, len(targetIndWithCommonGroupLabels))
+	var hashConflicts []commonlabels.LabelSet
+	for _, hash := range sortedKeys {
+		targetIndices := targetIndWithCommonGroupLabels[hash]
+		sharedLabels := tgs[targetIndices[0]].group // all should have the same group labels, so grab first one
```

Comment:
Is this enforced by the caller? If so since its an exported method this feels like it might not be always true.